Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean

```diff
@@ -757,6 +771,124 @@ def toConstProdLinearMap : (V1 →ᵃ[k] V2) ≃ₗ[R] V2 × (V1 →ₗ[k] V2) w
+  toFun := fun m a ↦ (f a).1 m
+  linear := LinearMap.pi (fun a ↦ (f a).2)
+  map_vadd' := fun _ _ ↦ funext fun a ↦ (f a).3 _ _
+
+--fp for when the image is a dependent AffineSpace φp i, fv for when the
+--image is a Module φv i, f' for when the image isn't dependent.
+variable (fp : (i : ι) → (P1 →ᵃ[k] φp i)) (fv : (i : ι) → (P1 →ᵃ[k] φv i))
+  (f' : ι → P1 →ᵃ[k] P2)
+
+@[simp]
+theorem pi_apply (c : P1) (i : ι) : pi fp c i = fp i c :=
+  rfl
+
+theorem pi_comp {V3 P3 : Type*} [AddCommGroup V3] [Module k V3] [AffineSpace V3 P3]
+    (g : P3 →ᵃ[k] P1) : (pi fp).comp g = pi (fun i => (fp i).comp g) :=
+  rfl
+
+theorem pi_eq_zero : pi fv = 0 ↔ ∀ i, fv i = 0 := by
+  simp only [AffineMap.ext_iff, Function.funext_iff, pi_apply]
+  exact forall_comm
+
+theorem pi_zero : pi (fun _ ↦ 0 : (i : ι) → P1 →ᵃ[k] φv i) = 0 := by
+  ext; rfl
+
+theorem proj_pi (i : ι) : (proj i).comp (pi fp) = fp i :=
+  ext fun _ => rfl
+
+/-- An ι-indexed collection of AffineMaps `P1 →ᵃ P2` lifts to an AffineMap from P1 to the
+  Module of functions, `ι → P2`. This module is `Pi.module`. This is the AffineMap analog
+  of `LinearMap.pi`. Actually defeq to `AffineMap.pi` but the type signature is
+  much easier to read.
+-/
+def pi' : (ι → (P1 →ᵃ[k] P2)) → P1 →ᵃ[k] (ι → P2) := pi
+
+@[simp]
+theorem pi'_apply (c : P1) (i : ι) : pi' f' c i = f' i c :=
+  rfl
+
+theorem pi'_comp {V3 P3 : Type*} [AddCommGroup V3] [Module k V3] [AffineSpace V3 P3]
+    (g : P3 →ᵃ[k] P1) : (pi' f').comp g = pi' (fun i => (f' i).comp g) :=
+  rfl
+
+theorem pi'_eq_zero (f : (i : ι) → P1 →ᵃ[k] V2) : pi' f = 0 ↔ ∀ i, f i = 0 :=
+  pi_eq_zero f
+
+theorem pi'_zero : pi' (fun _ ↦ 0 : (i : ι) → P1 →ᵃ[k] V2) = 0 := by ext; rfl
+
+theorem proj_pi' (i : ι) : (proj i).comp (pi' f') = f' i :=
+  ext fun _ => rfl
+
+/-- The `AffineMap` version of `LinearMap.single`, `AddMonoidHom.single` and `Pi.single`. -/
+def single [DecidableEq ι] (i : ι) : φv i →ᵃ[k] (j : ι) → φv j :=
+  { toFun := Pi.single i
+    linear := LinearMap.single i
+    map_vadd' := fun _ _ ↦ Pi.single_add i _ _
+  }
+
+@[simp]
+theorem coe_single [DecidableEq ι] (i : ι) :
+    ⇑(single i : φv i →ᵃ[k] (j : ι) → φv j) = Pi.single i :=
+  rfl
+
+section Ext
+
+variable [Finite ι] [DecidableEq ι] {f g : ((i : ι) → φv i) →ᵃ[k] P2}
+
+/- Two affine maps from a Pi-tyoe of modules `(i : ι) → φv i` are equal if they are equal in their
+  operation on `Pi.single` and at zero. -/
+theorem pi_ext_zero (h : ∀ i x, f (Pi.single i x) = g (Pi.single i x)) (h₂ : f 0 = g 0) :
+    f = g := by
+  apply ext_linear
+  next =>
+    apply LinearMap.pi_ext
+    intro i x
+    have s₁ := h i x
+    have s₂ := f.map_vadd 0 (Pi.single i x)
+    have s₃ := g.map_vadd 0 (Pi.single i x)
+    rw [vadd_eq_add, add_zero] at s₂ s₃
+    replace h₂ := h i 0
+    simp at h₂
+    rwa [s₂, s₃, h₂, vadd_right_cancel_iff] at s₁
+  next =>
+    exact h₂
+
+/- Two affine maps from a Pi-tyoe of modules `(i : ι) → φv i` are equal if they are equal in their
```

Comment:
```suggestion
/-- Two affine maps from a Pi-tyoe of modules `(i : ι) → φv i` are equal if they are equal in their
```
so that this shows up in the docs